posts/2024-05-07-ultimate-string-guide.md

```diff
@@ -0,0 +1,1250 @@
+
+## A word on lazy IO
+
+Some of the named packages expose API for reading and writing files via their lazy variants:
+
+- [Data.Text.Lazy.IO.readfile](https://hackage.haskell.org/package/text-2.1.1/docs/Data-Text-Lazy-IO.html#v:readFile)
+- [Data.ByteString.Lazy.readFile](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString-Lazy.html#v:readFile)
+
+Lazy IO is a hack to use incremental reading/processing without the use of a
+proper streaming library. The [bytestring documentation](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString-Lazy.html#g:25) warns about it:
+
+> * The program reads a file and writes the same file. This means that the file may be locked because the handler has not been released when writeFile is executed.
+> * The program reads thousands of files, but due to lazy evaluation, the OS's file descriptor limit is reached before the handlers can be released.
+
+Lazy IO makes it hard to reason about resources, order of execution etc.
+It is better to use a proper streaming library.
+
+## Streaming
+
+Streaming can not only solve the lazy IO problem, but may also
+solve some of the inefficiency of the `[Char]` type, while keeping a similarly
+simple API.
+
+There are many popular streaming libraries. A few of them are:
+
+- [conduit](https://hackage.haskell.org/package/conduit)
+- [streaming](https://hackage.haskell.org/package/streaming)
+- [streamly](https://hackage.haskell.org/package/streamly)
+- [pipes](https://hackage.haskell.org/package/pipes)
+
+### Via Streamly
+
+A couple of years ago I wrote the blog post
+[From conduit to streamly](https://hasufell.github.io/posts/2021-10-22-conduit-to-streamly.html),
+which gives an introduction into both streamly and conduit. The streamly API
+has diverged quite a bit since then, with multiple major versions. So I won't
+go into much detail about it.
+
+However, streamly is one notable example which provides an alternative
+to the `[Char]` type in [Streamly.Unicode.Stream](https://hackage.haskell.org/package/streamly-core-0.2.2/docs/Streamly-Unicode-Stream.html):
+
+```hs
+decodeUtf8 :: Monad m => Stream m Word8 -> Stream m Char
+encodeUtf8 :: Monad m => Stream m Char -> Stream m Word8
+```
+
+A very simple program to print the last unicode char of a file via streamly is:
+
+```hs
+import System.Environment (getArgs)
+import Streamly.Data.Stream (Stream)
+import Streamly.Data.Fold (Fold)
+import qualified Streamly.Data.Fold as Fold
+import qualified Streamly.FileSystem.File as File
+import qualified Streamly.Unicode.Stream as Unicode
+
+main :: IO ()
+main = do
+  (file:_) <- getArgs
+  c <- getFirstCharFromFile file
+  print c
+
+getFirstCharFromFile :: FilePath -> IO (Maybe Char)
+getFirstCharFromFile file = stream `Fold.drive` fold
+ where
+  stream :: Stream IO Char
+  stream = Unicode.decodeUtf8Chunks $ File.readChunks file
+
+  fold :: Monad m => Fold m a (Maybe a)
+  fold = Fold.one
```

Comment:
`Fold.latest`?
`streamly-string` uses the correct implementation although named `getFirstCharFromFile`.